zulipterminal/helper.py

```diff
@@ -459,15 +459,48 @@ def match_emoji(emoji: str, text: str) -> bool:
     return emoji.lower().startswith(text.lower())
 
 
-def match_stream(stream: Any, text: str) -> bool:
+def match_stream(streams: Dict[Any, List[str]], text: str,
+                 pinned_streams: List[str]) -> List[str]:
     """
-    True if the stream matches with `text` (case insensitive),
-    False otherwise.
+    Returns a list of streams whose words match with the 'text' in the
+    following order:
+    * 1st-word startswith match > 2nd-word startswith match > ... (pinned)
+    * 1st-word startswith match > 2nd-word startswith match > ... (unpinned)
+
+    Note: This function expects `streams` to be sorted in a non-decreasing
+    order.
     """
-    stream_name = stream[0].lower()
-    if stream_name.startswith(text.lower()):
-        return True
-    return False
+    pinned_stream_names = [stream[0] for stream in pinned_streams]
+
+    delimiters = [' ', '-', '_', '/']
+    for stream in streams:
+        stream_name = streams[stream][0]
+        for delimiter in delimiters:
+            if delimiter in stream_name:
+                streams[stream].extend(stream_name.split(delimiter)[1:])
+                break
+
+    matches = OrderedDict([
```

Comment:
For `matches`, an alternative could be to use a `list` (where index 0: 'pinned' and index 1: 'unpinned').

I went with an `OrderedDict` for 'readability' (see line 490).